src/kpca.jl

```diff
@@ -0,0 +1,162 @@
+        end
+    end
+    return K
+end
+
+"""Kernel PCA type"""
+immutable KernelPCA{T<:AbstractFloat}
+    X::AbstractMatrix{T}  # fitted data
+    ker::Function         # kernel function
+    center::KernelCenter  # kernel center
+    λ::DenseVector{T}     # eigenvalues  in feature space
+    α::DenseMatrix{T}     # eigenvectors in feature space
+    inv::DenseMatrix{T}   # inverse transform coefficients
+end
+
+## properties
+
+indim(M::KernelPCA) = size(M.X, 1)
+outdim(M::KernelPCA) = length(M.λ)
+
+projection(M::KernelPCA) = M.α ./ sqrt.(M.λ')
+principalvars(M::KernelPCA) = M.λ
+
+## use
+
+"""Calculate transformation to kernel space"""
+function transform{T<:AbstractFloat}(M::KernelPCA{T}, x::AbstractVecOrMat{T})
+    k = pairwise(M.ker, M.X, x)
+    transform!(M.center, k)
+    return projection(M)'*k
+end
+
+function transform{T<:AbstractFloat}(M::KernelPCA{T})
+    return projection(M)'*M.X
+end
+
+"""Calculate inverse transformation to original space"""
+function reconstruct{T<:AbstractFloat}(M::KernelPCA{T}, y::AbstractVecOrMat{T})
+    @assert size(M.inv, 1) > 0 "Inverse transformation coefficients are not available, set `inverse` parameter when fitting data"
+    Pᵗ = M.α' .* sqrt.(M.λ)
+    k = MultivariateStats.pairwise(M.ker, Pᵗ, y)
+    return M.inv*k
+end
+
+
+## show
+
+function Base.show(io::IO, M::KernelPCA)
+    print(io, "Kernel PCA(indim = $(indim(M)), outdim = $(outdim(M)))")
+end
+
+## core algorithms
+
+function pairwise!{T<:AbstractFloat}(K::AbstractVecOrMat{T}, kernel::Function,
+    X::AbstractVecOrMat{T}, Y::AbstractVecOrMat{T})
+    n = size(X, 2)
+    m = size(Y, 2)
+    for j = 1:m
+        aj = view(Y, :, j)
+        for i in j:n
+            @inbounds K[i, j] = kernel(view(X, :, i), aj)[]
+        end
+        j <= n && for i in 1:(j - 1)
+            @inbounds K[i, j] = K[j, i]   # leveraging the symmetry
+        end
+    end
+    K
+end
+
+pairwise!{T<:AbstractFloat}(K::AbstractVecOrMat{T}, kernel::Function, X::AbstractVecOrMat{T}) =
+    pairwise!(K, kernel, X, X)
+
+function pairwise{T<:AbstractFloat}(kernel::Function, X::AbstractVecOrMat{T}, Y::AbstractVecOrMat{T})
+    n = size(X, 2)
+    m = size(Y, 2)
+    K = similar(X, n, m)
+    pairwise!(K, kernel, X, Y)
+end
+
+pairwise{T<:AbstractFloat}(kernel::Function, X::AbstractVecOrMat{T}) =
+    pairwise(kernel, X, X)
+
+## interface functions
+
+function fit{T<:AbstractFloat}(::Type{KernelPCA}, X::AbstractMatrix{T};
+                               kernel = (x,y)->x'y,
+                               maxoutdim::Int = min(size(X)...),
+                               remove_zero_eig::Bool = false, atol::Real = 1e-10,
+                               solver::Symbol = :eig,
+                               inverse::Bool = false,  β::Real = 1.0,
+                               etol::Real = 0.0, etot::Real = 300)
+    d, n = size(X)
+    iskernelfunc = false
+
+    maxoutdim = min(min(d, n), maxoutdim)
+
+    K, Kfunc = if isa(kernel, Function)
```

Comment:
I would find it clearer to set `K` and `Kfunc` separately inside each branch. Same below.